unify Arduino devicetree specification for SPI-capable shields

pabigot opened the pull request:
Following up on a discussion in #21610 this PR unifies the minimal support for using SPI-based Arduino shields in each board devicetree that provides an `arduino_spi` label.  The changes are:

* Provide a `cs-gpios` property identifying the [standard](https://www.circuito.io/blog/arduino-uno-pinout/) Arduino Uno SPI chip select on D10 in all boards that provide `arduino_spi`;
* Replace GPIO specifiers in one shield that inappropriately used a specific board's GPIO controller directly;
* Remove the redundant `cs-gpios` properties from shields.

carlescufi reviewed (approved):
LGTM

jfischer-phytec-iot commented on boards/shields/link_board_can/link_board_can.overlay:
```diff
@@ -6,15 +6,12 @@
-	sck-pin = <36>;
-	mosi-pin = <37>;
-	miso-pin = <38>;
+	cs-gpios = <&arduino_header 16 0>; /* D10 */
```
I would not change it. This was a prototype, distributed to only one person, and it is not an Arduino connector shield. I will remove this shield soon as it is deprecated and the successor will be available after EW2020. Please drop this commit.

pabigot replied:
OK, changes removed.  This is very confusing: if this is not an Arduino-compatible shield, it shouldn't be referencing the `arduino_spi` label.  Perhaps this should just be removed since it's for an undocumented prototype.  If for some reason that can't be done now, please make sure its replacement uses a different connector name.

galak replied:
Agreed with @pabigot that we should either remove this is it was a prototype and not available to most people or at least clean it up to not use arduino references.

nandojve replied:
Why send a prototype support for the Zephyr Project? The internal prototype for one customer doesn't seems to belong there.

Under development SoC or like, if later will be renamed for the final product name works fine as the code is reviewed and a future renamed is effective but one customer prototype belongs to the customer fork.

jfischer-phytec-iot replied:
> Why send a prototype support for the Zephyr Project? The internal prototype for one customer doesn't seems to belong there.
> 
> Under development SoC or like, if later will be renamed for the final product name works fine as the code is reviewed and a future renamed is effective but one customer prototype belongs to the customer fork.

Support was inserted by a contributor. The product has been discontinued and will be replaced by a successor. What does it concern you and whats is the point of serving here?

alexanderwachter replied:
I submitted the shield as a demonstrator on how to integrate the MCP2515, since there is no other board with this specific controller on it. @karstenkoenig and I use this prototype to test the driver implementation. We kind of use it as the reference.

henrikbrixandersen commented:
Some of these boards may be using "native" SPI chip selects and therefore not enabling "D10" as a GPIO in their pinmux. Introducting `cs-gpios` for these boards will actually break Arduino SPI for those boards.

How much testing has gone into this?

pabigot commented:
It would be nice if there were some way to tell whether the bus used GPIO or internal chip select control (probably via pinmux), rather than relying on whether a driver populated a chip-select field.  Had that been readily available I wouldn't have updated all board files, only those that required it.

The SPI context appears to unconditionally manipulate CS GPIO regardless of whether the underlying hardware directly controls CS internally.  Peripherally related is https://github.com/zephyrproject-rtos/zephyr/issues/21611#issuecomment-575589246 which does something similar to the effect this might have.

Marked DNM until there's a clear way to identify SOCs that require/support manually-driven CS signals.   #12698 may be related.

pabigot committed:
boards: define Arduino SPI standard chip select for Nordic boards

Prepare for a day when all boards specify the chip select in the bus
devicetree node by making sure that the devicetree node associated
with arduino_spi uses the standard D10 pin from the arduino header as
the cs-gpios property value.

Also update the arduino tags in the board yaml files.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Ok, this is a mess that's out of scope for me.  [This comment](https://github.com/zephyrproject-rtos/zephyr/issues/12226#issuecomment-452888111) says we're supposed to use GPIO specification for CS consistently, but that doesn't appear to be what's actually supported.

It may or may not be OK to have `cs-gpios` specified for a platform that uses only hardware-indexed chip selects.  Depending on how pinmux ultimately manages functions (e.g. does `gpio_pin_configure()` override a previous pinmux assignment to a non-GPIO function; @galak @mnkp please make sure this is defined in pinmux requirements) this could work, or it could be ignored.  I wouldn't expect it to cause an error normally.  The original approach taken in this PR does break several boards that define `arduino_spi` but not `arduino_header`.

Also boards that do use hardware-indexed pins rather than GPIOs may or may not allow GPIO control instead/alongside, based on Kconfig.  E.g. `CONFIG_SPI_STM32_USE_HW_SS=y` appears to switch between peripheral-defined and GPIO control at the SOC level.  This makes it more difficult to figure out how to specify the chip select generically.

I have limited SPI devices on breakouts to try this (none in Arduino form-factor), and the one that works on nrf51_ble400 for SDHC does not work on frdm_k64f in any configuration I can come up with.

So: I've reduced this to simply adding the cs-gpios consistently on all Nordic boards, leaving the improper defines in the shields where they do not belong.  Baby steps....

vanwinkeljan commented:
@pabigot Could you also update the nordic boards yaml files to include `arduino_spi`?

pabigot commented:
@vanwinkeljan done.  We should probably have some sort of check for this, if those tags are important to testing.

pabigot commented:
Sanity check failure due to #21568, open for four weeks with no assignee.  IMO this should be mergeable.
